onnxruntime/python/tools/transformers/fusion_rotary_attention.py

```diff
@@ -885,21 +921,108 @@ def fuse(self, normalize_node, input_name_to_nodes, output_name_to_node):
+            hidden_size_concat_node = get_num_heads_and_hidden_size(reshape_k)
+
+            transpose_node_name = self.model.create_node_name("Transpose")
+
+            tranpose_output_name = transpose_node_name + "_output_0"
+
+            transpose_node = helper.make_node("Transpose", inputs=[concat_k_half.output[0]], outputs=[tranpose_output_name], name=transpose_node_name)
+            transpose_node.attribute.extend([helper.make_attribute("perm", [0, 2, 1, 3])])
+
+            # Reshape the Rotary Embedding output for key
+            concat_k_reshape_node_name = self.model.create_node_name("Reshape", name_prefix="concat_k_half")
+            concat_k_reshape_node = helper.make_node(
+                "Reshape",
+                inputs=[transpose_node.output[0], hidden_size_concat_node.output[0]],
+                outputs=[concat_k_reshape_node_name + "_output_0"],
+                name=concat_k_reshape_node_name,
+            )
+
+            q_transpose_node_name = self.model.create_node_name("Transpose")
+
+            q_tranpose_output_name = q_transpose_node_name + "_output_0"
+
+            q_transpose_node = helper.make_node("Transpose", inputs=[concat_q_half.output[0]], outputs=[q_tranpose_output_name],
+                                            name=q_transpose_node_name)
+            q_transpose_node.attribute.extend([helper.make_attribute("perm", [0, 2, 1, 3])])
+            # Reshape the Rotary Embedding output for query
+            concat_q_reshape_node_name = self.model.create_node_name("Reshape", name_prefix="concat_q_half")
+
+            concat_q_reshape_node = helper.make_node(
+                "Reshape",
+                inputs=[q_transpose_node.output[0], hidden_size_concat_node.output[0]],
+                outputs=[concat_q_reshape_node_name + "_output_0"],
+                name=concat_q_reshape_node_name,
+            )
+            self.nodes_to_add.append(hidden_size_concat_node)
+            self.nodes_to_add.append(transpose_node)
+            self.nodes_to_add.append(q_transpose_node)
+            self.nodes_to_add.append(concat_k_reshape_node)
+            self.nodes_to_add.append(concat_q_reshape_node)
+
+            self.node_name_to_graph_name[hidden_size_concat_node.name] = self.this_graph_name
+            self.node_name_to_graph_name[transpose_node.name] = self.this_graph_name
+            self.node_name_to_graph_name[q_transpose_node.name] = self.this_graph_name
+            self.node_name_to_graph_name[concat_k_reshape_node.name] = self.this_graph_name
+            self.node_name_to_graph_name[concat_q_reshape_node.name] = self.this_graph_name
+
         new_node = self.create_mha_node(
             matmul_q.input[0],
             root_output,
-            rotary_q,
-            rotary_k,
-            matmul_v,
+            concat_q_reshape_node if concat_q_half else rotary_q,
+            concat_k_reshape_node if concat_k_half else rotary_k,
+            add_v if add_v else matmul_v,
```

Comment:
Instead of changing this, can we assign `rotary_q`, `rotary_k` and `matmul_v` to `concat_q_reshape_node`, `concat_k_reshape_node` and `add_v` before that line if they are not `None`? It will make it clearer to the onnxruntime team later that we don't break anything on their side.